src/nameresolution/mod.rs

```diff
@@ -1627,7 +1627,10 @@ impl<'c> Resolvable<'c> for ast::Handle<'c> {
 impl<'c> Resolvable<'c> for ast::NamedConstructor<'c> {
     fn declare(&mut self, _resolver: &mut NameResolver, _cache: &mut ModuleCache<'c>) {}
 
-    fn define(&mut self, _resolver: &mut NameResolver, _cache: &mut ModuleCache<'c>) {
-        todo!()
+    fn define(&mut self, resolver: &mut NameResolver, cache: &mut ModuleCache<'c>) {
+        self.constructor.define(resolver, cache);
+        for (_, arg_val) in self.args.iter_mut() {
+            arg_val.define(resolver, cache);
+        }
```

Comment:
I think you'd save a lot of code if you desugared named constructors into ordinary constructors here. You'd still need to check all fields are present though.

It should desugar to roughly:
```
Sequence(vec![
    Definition(new_var1, field_expr1),
    ...
    Definition(new_varN, field_exprN),
    FunctionCall(Variable(TypeConstructor(Foo)), vec![new_var1, ..., new_varN])
])
```
(where the fields are defined in the order they were specified in the named constructor expr, but applied to the function call in the order Foo's definition expects)
That way we'd get type checking + monomorphization for free

Reply:
Do you mean desugaring into a sequence in the parsing stage, or somehow mutating `self` into a `sequence` in the name resolution stage? Cause I was not certain if we would even have the type info of the struct we're trying to construct at the name resolution stage yet, and as such we would not be able to call the constructor in the right order.

Reply:
I mean mutating `self` during `NameResolution::define` of a named constructor.

Looking at `TypeDefinition::define` though, I suppose we wouldn't necessarily know a types fields until that is called. I suppose we could also do the same mutation but during type checking. Then recursively type check the result.

Reply:
I've decided to just store the inner `Sequence` as a field on `NamedConstructor`, and mutate it during the `define` phase.

Reply:
That's probably a better idea. It'd be tricky trying to mutate a named constructor into a sequence without acces to the actual `&mut Ast` value.